File: packages/core/tools/matimo_list_user_tools/matimo_list_user_tools.ts
import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import {
  validateToolDefinition,
  classifyRisk,
  getGlobalMatimoLogger,
} from '@matimo/core';

interface ListParams {
  tool_dir?: string;
  include_drafts?: boolean;
}

interface ToolSummary {
  name: string;
  description: string;
  version: string;
  status: string;
  riskLevel: string;
  tags: string[];
}

interface ListResult {
  tools: ToolSummary[];
  total: number;
}

export default async function matimoListUserTools(
  params: ListParams,
): Promise<ListResult> {
  const logger = getGlobalMatimoLogger();
  const toolDir = params.tool_dir || './matimo-tools';
  const includeDrafts = params.include_drafts !== false;

  const tools: ToolSummary[] = [];

  if (!fs.existsSync(toolDir)) {
    return { tools: [], total: 0 };
  }

  const entries = fs.readdirSync(toolDir, { withFileTypes: true });
  for (const entry of entries) {
    if (!entry.isDirectory()) continue;

    const defPath = path.join(toolDir, entry.name, 'definition.yaml');
    if (!fs.existsSync(defPath)) continue;

    try {
      const content = fs.readFileSync(defPath, 'utf-8');
      const parsed = yaml.load(content);
      const tool = validateToolDefinition(parsed);

      const status = tool.status || 'approved';
      if (!includeDrafts && status === 'draft') continue;

      tools.push({
        name: tool.name,
        description: tool.description,
        version: tool.version,
        status,
        riskLevel: classifyRisk(tool),
        tags: tool.tags || [],
      });
    } catch (err) {
      logger.warn('matimo_list_user_tools: failed to parse tool', {
        dir: entry.name,
        error: (err as Error).message,
      });
    }
  }

  return { tools, total: tools.length };
}

File: python/packages/core/src/matimo/tools/matimo_list_user_tools/matimo_list_user_tools.py
"""matimo_list_user_tools — list all user-created tools in a directory."""
from __future__ import annotations

import logging
from pathlib import Path

import yaml

logger = logging.getLogger("matimo")


async def run(params: dict) -> dict:  # type: ignore[type-arg]
    from matimo.core.models import ToolDefinition
    from matimo.policy.risk_classifier import classify_risk

    tool_dir: str = params.get("tool_dir", "./matimo-tools")
    include_drafts: bool = bool(params.get("include_drafts", True))

    root = Path(tool_dir)
    if not root.exists():
        return {"tools": [], "total": 0}

    tools = []
    for subdir in sorted(root.iterdir()):
        if not subdir.is_dir():
            continue
        def_path = subdir / "definition.yaml"
        if not def_path.exists():
            continue
        try:
            parsed = yaml.safe_load(def_path.read_text(encoding="utf-8"))
            tool_def = ToolDefinition.model_validate(parsed)
            status = str(tool_def.status)
            if not include_drafts and status == "draft":
                continue
            try:
                risk_level = classify_risk(tool_def)
            except Exception:
                risk_level = "medium"
            tools.append({
                "name": tool_def.name,
                "description": tool_def.description,
                "version": tool_def.version,
                "status": status,
                "riskLevel": risk_level,
                "tags": tool_def.tags or [],
            })
        except Exception as exc:
            logger.warning("matimo_list_user_tools: failed to parse %s: %s", def_path, exc)

    return {"tools": tools, "total": len(tools)}
